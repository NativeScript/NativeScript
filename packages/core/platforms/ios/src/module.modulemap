module NativeScriptEmbedder {
    header "NativeScriptEmbedder.h"
    export *
}

module NativeScriptUtils {
    header "NativeScriptUtils.h"
    export *
}

module UIViewNativeScript {
    header "UIView+NativeScript.h"
    export *
}
module TNSWidgets {
    umbrella header "TNSWidgets/TNSWidgets.h"
    module * { export * }
    export *
}