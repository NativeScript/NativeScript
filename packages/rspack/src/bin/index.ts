#!/usr/bin/env node

import { Configuration, rspack, Stats } from '@rspack/core';
import { RspackError } from '@rspack/core/dist/RspackError';
import { green, greenBright, redBright, yellow } from 'ansi-colors';
import { program } from 'commander';
import fs from 'fs';
import path from 'path';
import dedent from 'ts-dedent';
import { parseEnvFlags } from '../cli/parseEnvFlags';

const defaultConfig = path.resolve(
	__dirname,
	'../stubs/default.config.stub.js',
);
const tag = `[${green('../index')}]`;

function error(message: string) {
	console.error(`${tag} ${redBright(dedent(message))}`);
}

function info(message: string) {
	console.info(`${tag} ${greenBright(dedent(message))}`);
}

program.enablePositionalOptions();

program
	.command('init')
	.description('Initialize a new webpack.config.js in the current directory.')
	.action(() => {
		const targetPath = path.resolve(process.cwd(), 'webpack.config.js');

		if (fs.existsSync(targetPath)) {
			return error(`File Already Exists: ${targetPath}`);
		}

		fs.copyFileSync(defaultConfig, targetPath);

		info('Initialized config.');
	});

program
	.command('build')
	.description('Build...')
	.option('--env [name]', 'environment name')
	.option('--config [path]', 'config path')
	.option('--watch', 'watch for changes')
	.allowUnknownOption()
	.action((options, command) => {
		const env = parseEnvFlags(command.args);
		// add --env <val> into the env object
		// for example if we use --env prod
		// we'd have env.env = 'prod'
		if (options.env) {
			env['env'] = options.env;
		}

		env['stats'] ??= true;
		env['watch'] ??= options.watch;

		// if --env.config is passed, we'll set an environment
		// variable to it's value so that the config Util
		// reads from the correct config file.
		process.env.NATIVESCRIPT_CONFIG_NAME ??= env['config'];

		const configPath = (() => {
			if (options.config) {
				return path.resolve(options.config);
			}

			return path.resolve(process.cwd(), 'webpack.config.js');
		})();

		// todo: validate config exists
		// todo: guard against invalid config
		let configuration: Configuration;
		try {
			configuration = require(configPath)(env);
		} catch (err) {
			console.log(err);
		}

		if (!configuration) {
			console.log('No configuration!');
			process.exitCode = 1;
			return;
		}

		const compiler = rspack(configuration);

		const webpackCompilationCallback = (err: RspackError, stats: Stats) => {
			if (err) {
				// Do not keep cache anymore
				compiler.purgeInputFileSystem();

				console.error(err.stack || err);
				if (err.message) {
					console.error(err.message);
				}

				process.exitCode = 1;
				return;
			}

			if (stats) {
				// Set the process exit code depending on errors
				process.exitCode = stats.hasErrors() ? 1 : 0;

				if (env.stats) {
					console.log(
						stats.toString({
							chunks: false,
							colors: true,
							errorDetails: env.verbose,
						}),
					);
				}

				// if webpack profile is enabled we write the stats to a JSON file
				if (configuration.profile || env.profile) {
					console.log(
						[
							'',
							'|',
							`|  The build profile has been written to ${yellow(
								'webpack.stats.json',
							)}`,
							`|  You can analyse the stats at ${green(
								'https://webpack.github.io/analyse/',
							)}`,
							'|',
							'',
						].join('\n'),
					);
					fs.writeFileSync(
						path.join(process.cwd(), 'webpack.stats.json'),
						JSON.stringify(stats.toJson()),
					);
				}
			}
		};

		if (options.watch) {
			env.stats && console.log('rspack is watching the files...');
			compiler.watch(
				configuration.watchOptions ?? {},
				webpackCompilationCallback,
			);
		} else {
			compiler.run((err, stats) => {
				compiler.close((err2) =>
					webpackCompilationCallback((err || err2) as RspackError, stats),
				);
			});
		}
	});

program.version(require('../../package.json').version, '-v, --version');

program.parse(process.argv);
