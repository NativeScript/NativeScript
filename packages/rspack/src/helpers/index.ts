import { merge } from 'webpack-merge';

import {
	getPackageJson,
	getProjectRootPath,
	getProjectFilePath,
} from './project';
import { addVirtualEntry, addVirtualModule } from './virtualModules';
import { applyFileReplacements } from './fileReplacements';
import { addCopyRule, removeCopyRule } from './copyRules';
import { error, info, warn, warnOnce } from './log';
import { determineProjectFlavor } from './flavor';
import { readTsConfig } from './typescript';
import { getValue } from './config';
import { getIPS } from './host';
import {
	getAllDependencies,
	hasDependency,
	getDependencyPath,
} from './dependencies';
import {
	addPlatform,
	getAbsoluteDistPath,
	getDistPath,
	getEntryDirPath,
	getEntryPath,
	getPlatform,
	getPlatformName,
} from './platform';

// intentionally populated manually
// as this generates nicer typings
// that show all the utils inline
// rather than imports to types
// todo: maybe use api-extractor instead
export default {
	merge,
	addCopyRule,
	removeCopyRule,
	applyFileReplacements,
	config: {
		getValue,
	},
	dependencies: {
		getAllDependencies,
		hasDependency,
		getDependencyPath,
	},
	flavor: {
		determineProjectFlavor,
	},
	host: {
		getIPS,
	},
	log: {
		error,
		info,
		warn,
		warnOnce,
	},
	platform: {
		addPlatform,
		getAbsoluteDistPath,
		getDistPath,
		getEntryDirPath,
		getEntryPath,
		getPlatform,
		getPlatformName,
	},
	project: {
		getProjectFilePath,
		getProjectRootPath,
		getPackageJson,
	},
	virtualModules: {
		addVirtualEntry,
		addVirtualModule,
	},
	tsconfig: {
		readTsConfig,
	},
};
