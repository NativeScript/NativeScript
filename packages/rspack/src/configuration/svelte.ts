import Config from 'rspack-chain';

import { hasDependency } from '../helpers/dependencies';
import { error } from '../helpers/log';
import { getPlatformName } from '../helpers/platform';
import { getProjectFilePath } from '../helpers/project';
import { env as _env, IWebpackEnv } from '../index';
import base from './base';

export default function (config: Config, env: IWebpackEnv = _env): Config {
	base(config, env);

	const platform = getPlatformName();
	const mode = env.production ? 'production' : 'development';
	const production = mode === 'production';

	// target('node') is the default but causes svelte-loader to detect it as a "server" render, disabling HMR
	// electron-main sneaks us past the target == 'node' check and gets us HMR
	config.target('electron-main');

	// turns out this isn't enough now. svelte uses "node" of which "electron-main" is a subset in its export map forcing imports
	// for 'svelte' to 'ssr.mjs'. We define an alias here to force it back.
	config.resolve.alias.set('svelte$', 'svelte/internal');

	// svelte-hmr still references tns-core-modules, so we shim it here for compat.
	config.resolve.alias.set('tns-core-modules', '@nativescript/core');

	// resolve .svelte files
	// the order is reversed because we are using prepend!
	config.resolve.extensions.prepend('.svelte').prepend(`.${platform}.svelte`);

	// add worker support to .svelte files (new Worker('./path'))
	config.module.rule('workers').test(/\.(js|ts|svelte)$/);

	// add a rule for .svelte files
	if (hasDependency('svelte-loader')) {
		// configure using svelte-loader
		config.module
			.rule('svelte')
			.test(/\.svelte$/)
			.exclude.add(/node_modules/)
			.end()
			.use('svelte-loader')
			.loader('svelte-loader')
			.tap((options) => {
				const svelteConfig = getSvelteConfig();
				return {
					...options,
					compilerOptions: {
						...svelteConfig?.compilerOptions,
						dev: !production,
					},
					preprocess: svelteConfig?.preprocess,
					hotReload: !production,
					hotOptions: {
						injectCss: false,
						native: true,
					},
				};
			});
	} else {
		// fallback for projects still using svelte-loader-hot
		config.module
			.rule('svelte')
			.test(/\.svelte$/)
			.exclude.add(/node_modules/)
			.end()
			.use('svelte-loader-hot')
			.loader('svelte-loader-hot')
			.tap((options) => {
				const svelteConfig = getSvelteConfig();
				return {
					...options,
					dev: !production,
					preprocess: getSvelteConfigPreprocessor(),
					hotReload: !production,
					hotOptions: {
						injectCss: false,
						native: true,
					},
					// Suppress A11y warnings
					onwarn(warning, warn) {
						if (!/A11y:/.test(warning.message)) {
							warn(warning);
						}
					},
				};
			});
	}

	// todo: re-visit later, disabling by default now
	// config.plugin('DefinePlugin').tap((args) => {
	// 	args[0] = merge(args[0], {
	// 		__UI_USE_EXTERNAL_RENDERER__: true,
	// 	});

	// 	return args;
	// });

	return config;
}

function getSvelteConfigPreprocessor(): any {
	const config = getSvelteConfig();

	return config?.preprocess;
}

interface ISvelteConfig {
	preprocess: any;
	compilerOptions: any;
}

function getSvelteConfig(): ISvelteConfig | undefined {
	try {
		return require(getProjectFilePath('svelte.config.js')) as ISvelteConfig;
	} catch (err) {
		error('Could not find svelte.config.js.', err);
	}
}
