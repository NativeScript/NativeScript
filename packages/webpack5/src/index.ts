// Make sure the Acorn Parser (used by Webpack) can parse ES-Stage3 code
// This must be at the top BEFORE webpack is loaded so that we can extend
// and replace the parser before webpack uses it
// Based on the issue: https://github.com/webpack/webpack/issues/10216
import stage3 from 'acorn-stage3';

// we use require to be able to override the exports
const acorn = require('acorn');
acorn.Parser = acorn.Parser.extend(stage3);

import { highlight } from 'cli-highlight';
import { merge } from 'webpack-merge';
import Config from 'webpack-chain';
import webpack from 'webpack';

import { applyExternalConfigs } from './helpers/externalConfigs';
import { determineProjectFlavor } from './helpers/flavor';
import { error, info } from './helpers/log';
import { configs } from './configuration';
import helpers from './helpers';

export interface IWebpackEnv {
	[name: string]: any;

	env?: string;

	appPath?: string;
	appResourcesPath?: string;
	buildPath?: string;
	appComponents?: string[];

	nativescriptLibPath?: string | boolean;

	android?: boolean;
	ios?: boolean;
	// for custom platforms
	platform?: string;

	sourceMap?: string | boolean;
	production?: boolean;
	report?: boolean;
	hmr?: boolean;

	// enable verbose output
	verbose?: boolean;

	// enable webpack profiling
	profile?: boolean;

	// print webpack stats (default: true)
	stats?: boolean;

	// enable commonjs modules (default: ES modules, esm)
	commonjs?: boolean;

	// misc
	replace?: string[] | string;
	watchNodeModules?: boolean;
}

interface IChainEntry {
	chainFn: any;
	order?: number;
	plugin?: string;
}

let webpackChains: IChainEntry[] = [];
let webpackMerges: any[] = [];
let explicitUseConfig = false;
let hasInitialized = false;
let currentPlugin: string | undefined;

/**
 * @internal
 */
export let env: IWebpackEnv = {};

/**
 * @internal
 */
export function setCurrentPlugin(plugin: string) {
	currentPlugin = plugin;
}

/**
 * @internal
 */
export function clearCurrentPlugin() {
	currentPlugin = undefined;
}

////// PUBLIC API

/**
 * The default flavor specific configs
 */
export const defaultConfigs = configs;

/**
 * Utilities to simplify various tasks
 */
export const Utils = helpers;

/**
 * webpack-merge exported for convenience. Useful for merging configuration objects
 */
export { merge };

/**
 * Initialize @nativescript/webpack with the webpack env.
 * Must be called first.
 *
 * @param _env The webpack env
 */
export function init(_env: IWebpackEnv) {
	hasInitialized = true;
	if (_env) {
		env = _env;
	}
}

/**
 * Explicitly specify the base config to use.
 * Calling this will opt-out from automatic flavor detection.
 *
 * Useful when the flavor cannot be detected due to the project structure
 * for example in a custom monorepo.
 *
 * @param config Name of the base config to use.
 */
export function useConfig(config: keyof typeof defaultConfigs | false) {
	explicitUseConfig = true;
	if (config) {
		webpackChains.push({
			order: -1,
			chainFn: configs[config],
		});
	}
}

/**
 * Add a new function to be called when building the internal config using webpack-chain.
 *
 * @param chainFn A function that accepts the internal chain config, and the current environment
 * @param options Optional options to control the order in which the chain function should be applied.
 */
export function chainWebpack(
	chainFn: (config: Config, env: IWebpackEnv) => any,
	options?: { order?: number },
) {
	webpackChains.push({
		order: options?.order || 0,
		chainFn,
		plugin: currentPlugin,
	});
}

/**
 * Merge an object into the resolved chain config.
 *
 * @param mergeFn An object or a function that optionally returns an object (can mutate the object directly and return nothing)
 */
export function mergeWebpack(
	mergeFn:
		| ((config: Partial<webpack.Configuration>, env: IWebpackEnv) => any)
		| Partial<webpack.Configuration>,
) {
	webpackMerges.push(mergeFn);
}

/**
 * Resolve a new instance of the internal chain config with all chain functions applied.
 */
export function resolveChainableConfig(): Config {
	const config = new Config();

	if (!explicitUseConfig) {
		useConfig(determineProjectFlavor());
	}

	// apply configs from dependencies
	// todo: allow opt-out
	applyExternalConfigs();

	webpackChains
		.splice(0)
		.sort((a, b) => {
			return a.order - b.order;
		})
		.forEach(({ chainFn, plugin }) => {
			try {
				chainFn(config, env);
			} catch (err) {
				if (plugin) {
					// catch and print errors from plugins
					return error(`
						Unable to apply chain function from: ${plugin}.
						Error is: ${err}
					`);
				}

				// otherwise throw - as the error is likely from the user config
				// or missing env flags (eg. missing platform)
				throw err;
			}
		});

	if (env.verbose) {
		info('Resolved chainable config (before merges):');
		info(highlight(config.toString(), { language: 'js' }));
	}

	return config;
}

/**
 * Resolve a "final" configuration that has all chain functions and merges applied.
 *
 * @param chainableConfig Optional chain config to use.
 */
export function resolveConfig(
	chainableConfig = resolveChainableConfig(),
): webpack.Configuration {
	if (!hasInitialized) {
		throw error('resolveConfig() must be called after init()');
	}

	let config = chainableConfig.toConfig();

	// this applies webpack merges
	webpackMerges.forEach((mergeFn) => {
		if (typeof mergeFn === 'function') {
			// mergeFn is a function with optional return value
			const res = mergeFn(config, env);
			if (res) config = merge(config, res);
		} else if (mergeFn) {
			// mergeFn is a literal value (object)
			config = merge(config, mergeFn);
		}
	});

	// return a config usable by webpack
	return config;
}
