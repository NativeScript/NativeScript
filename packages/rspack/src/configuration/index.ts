import base from './base';

import angular from './angular';
import javascript from './javascript';
import react from './react';
import svelte from './svelte';
import typescript from './typescript';
import vue from './vue';
import vue3 from './vue3';

export const configs = {
	base,
	angular,
	javascript,
	react,
	svelte,
	typescript,
	vue,
	vue3,
};
