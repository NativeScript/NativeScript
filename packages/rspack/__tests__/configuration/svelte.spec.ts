import Config from 'webpack-chain';

import svelte from '../../src/configuration/svelte';
import { init } from '../../src';

jest.mock('__jest__/svelte.config.js', () => {}, { virtual: true });

describe('svelte configuration', () => {
	const platforms = ['ios', 'android'];

	for (let platform of platforms) {
		it(`for ${platform}`, () => {
			init({
				[platform]: true,
			});
			expect(svelte(new Config()).toString()).toMatchSnapshot();
		});
	}
});
