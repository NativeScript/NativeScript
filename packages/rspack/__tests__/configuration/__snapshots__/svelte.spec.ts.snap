// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`svelte configuration for android 1`] = `
"{
  mode: 'development',
  externals: [
    'package.json',
    '~/package.json'
  ],
  externalsPresets: {
    node: false
  },
  devtool: 'inline-source-map',
  target: 'electron-main',
  watchOptions: {
    ignored: [
      '__jest__/platforms/**',
      '__jest__/App_Resources/**'
    ]
  },
  ignoreWarnings: [
    /System.import\\(\\) is deprecated/
  ],
  output: {
    path: '__jest__/platforms/android/app/src/main/assets/app',
    pathinfo: false,
    publicPath: '',
    libraryTarget: 'commonjs',
    globalObject: 'global',
    clean: true
  },
  resolve: {
    symlinks: true,
    alias: {
      '~': '__jest__/src',
      '@': '__jest__/src',
      svelte$: 'svelte/internal',
      'tns-core-modules': '@nativescript/core'
    },
    extensions: [
      '.android.svelte',
      '.svelte',
      '.android.ts',
      '.ts',
      '.android.js',
      '.js',
      '.android.mjs',
      '.mjs',
      '.android.css',
      '.css',
      '.android.scss',
      '.scss',
      '.android.json',
      '.json'
    ],
    modules: [
      '__jest__/node_modules',
      'node_modules'
    ]
  },
  resolveLoader: {
    modules: [
      '__jest__/node_modules/@nativescript/webpack/dist/loaders',
      '__jest__/node_modules',
      'node_modules'
    ]
  },
  module: {
    rules: [
      /* config.module.rule('bundle') */
      {
        enforce: 'post',
        test: '__jest__/src/app.js',
        use: [
          /* config.module.rule('bundle').use('app-css-loader') */
          {
            loader: 'app-css-loader',
            options: {
              platform: 'android'
            }
          }
        ]
      },
      /* config.module.rule('workers') */
      {
        test: /\\.(js|ts|svelte)$/,
        use: [
          /* config.module.rule('workers').use('nativescript-worker-loader') */
          {
            loader: 'nativescript-worker-loader'
          }
        ]
      },
      /* config.module.rule('ts') */
      {
        test: [
          /\\.ts$/
        ],
        use: [
          /* config.module.rule('ts').use('ts-loader') */
          {
            loader: 'ts-loader',
            options: {
              transpileOnly: true,
              allowTsInNodeModules: true,
              compilerOptions: {
                sourceMap: true,
                declaration: false
              },
              getCustomTransformers: function () { /* omitted long function */ }
            }
          }
        ]
      },
      /* config.module.rule('js') */
      {
        test: /\\.js$/,
        exclude: [
          /node_modules/
        ]
      },
      /* config.module.rule('xml') */
      {
        test: /\\.xml$/,
        use: [
          /* config.module.rule('xml').use('xml-namespace-loader') */
          {
            loader: 'xml-namespace-loader'
          }
        ]
      },
      /* config.module.rule('css') */
      {
        test: /\\.css$/,
        use: [
          /* config.module.rule('css').use('apply-css-loader') */
          {
            loader: 'apply-css-loader'
          },
          /* config.module.rule('css').use('css2json-loader') */
          {
            loader: 'css2json-loader'
          },
          /* config.module.rule('css').use('postcss-loader') */
          {
            loader: 'postcss-loader',
            options: {
              postcssOptions: {
                plugins: [
                  [
                    'postcss-import',
                    {
                      resolve: function () { /* omitted long function */ }
                    }
                  ]
                ]
              }
            }
          }
        ]
      },
      /* config.module.rule('scss') */
      {
        test: /\\.scss$/,
        use: [
          /* config.module.rule('scss').use('apply-css-loader') */
          {
            loader: 'apply-css-loader'
          },
          /* config.module.rule('scss').use('css2json-loader') */
          {
            loader: 'css2json-loader'
          },
          /* config.module.rule('scss').use('postcss-loader') */
          {
            loader: 'postcss-loader',
            options: {
              postcssOptions: {
                plugins: [
                  [
                    'postcss-import',
                    {
                      resolve: function () { /* omitted long function */ }
                    }
                  ]
                ]
              }
            }
          },
          /* config.module.rule('scss').use('sass-loader') */
          {
            loader: 'sass-loader'
          }
        ]
      },
      /* config.module.rule('svelte') */
      {
        test: /\\.svelte$/,
        exclude: [
          /node_modules/
        ],
        use: [
          /* config.module.rule('svelte').use('svelte-loader-hot') */
          {
            loader: 'svelte-loader-hot',
            options: {
              dev: true,
              preprocess: undefined,
              hotReload: true,
              hotOptions: {
                injectCss: false,
                'native': true
              },
              onwarn: function () { /* omitted long function */ }
            }
          }
        ]
      }
    ]
  },
  optimization: {
    runtimeChunk: 'single',
    splitChunks: {
      cacheGroups: {
        defaultVendor: {
          test: /[\\\\/]node_modules[\\\\/]/,
          priority: -10,
          name: 'vendor',
          chunks: 'all'
        }
      }
    },
    minimizer: [
      /* config.optimization.minimizer('TerserPlugin') */
      new TerserPlugin(
        {
          terserOptions: {
            compress: {
              collapse_vars: false,
              sequences: false,
              keep_infinity: true,
              drop_console: false,
              global_defs: {
                __UGLIFIED__: true
              }
            },
            keep_fnames: true,
            keep_classnames: true,
            format: {
              keep_quoted_props: true
            }
          }
        }
      )
    ]
  },
  plugins: [
    /* config.plugin('ForkTsCheckerWebpackPlugin') */
    new ForkTsCheckerWebpackPlugin(
      {
        async: false,
        typescript: {
          memoryLimit: 4096
        }
      }
    ),
    /* config.plugin('PlatformSuffixPlugin') */
    new PlatformSuffixPlugin(
      {
        extensions: [
          'android'
        ]
      }
    ),
    /* config.plugin('ContextExclusionPlugin|App_Resources') */
    new ContextExclusionPlugin(
      /(.*)App_Resources(.*)/
    ),
    /* config.plugin('ContextExclusionPlugin|Other_Platforms') */
    new ContextExclusionPlugin(
      /\\.(ios|visionos)\\.(\\w+)$/
    ),
    /* config.plugin('DefinePlugin') */
    new DefinePlugin(
      {
        __DEV__: true,
        __NS_WEBPACK__: true,
        __NS_ENV_VERBOSE__: false,
        __NS_DEV_HOST_IPS__: '["127.0.0.1","192.168.0.10"]',
        __CSS_PARSER__: '"css-tree"',
        __UI_USE_XML_PARSER__: true,
        __UI_USE_EXTERNAL_RENDERER__: false,
        __ANDROID__: true,
        __IOS__: false,
        __VISIONOS__: false,
        __APPLE__: false,
        'global.isAndroid': true,
        'global.isIOS': false,
        'global.isVisionOS': false,
        process: 'global.process'
      }
    ),
    /* config.plugin('CopyWebpackPlugin') */
    new CopyPlugin(
      {
        patterns: [
          {
            from: 'assets/**',
            context: '__jest__/src',
            noErrorOnMissing: true,
            globOptions: {
              dot: false,
              ignore: []
            }
          },
          {
            from: 'fonts/**',
            context: '__jest__/src',
            noErrorOnMissing: true,
            globOptions: {
              dot: false,
              ignore: []
            }
          },
          {
            from: '**/*.+(jpg|png)',
            context: '__jest__/src',
            noErrorOnMissing: true,
            globOptions: {
              dot: false,
              ignore: []
            }
          }
        ]
      }
    ),
    /* config.plugin('WatchStatePlugin') */
    new WatchStatePlugin()
  ],
  entry: {
    bundle: [
      '@nativescript/core/globals/index',
      '@nativescript/core/bundle-entry-points',
      '__jest__/src/app.js',
      '@nativescript/core/ui/frame',
      '@nativescript/core/ui/frame/activity'
    ]
  }
}"
`;

exports[`svelte configuration for ios 1`] = `
"{
  mode: 'development',
  externals: [
    'package.json',
    '~/package.json'
  ],
  externalsPresets: {
    node: false
  },
  devtool: 'inline-source-map',
  target: 'electron-main',
  watchOptions: {
    ignored: [
      '__jest__/platforms/**',
      '__jest__/App_Resources/**'
    ]
  },
  ignoreWarnings: [
    /System.import\\(\\) is deprecated/
  ],
  output: {
    path: '__jest__/platforms/ios/jest/app',
    pathinfo: false,
    publicPath: '',
    libraryTarget: 'commonjs',
    globalObject: 'global',
    clean: true
  },
  resolve: {
    symlinks: true,
    alias: {
      '~': '__jest__/src',
      '@': '__jest__/src',
      svelte$: 'svelte/internal',
      'tns-core-modules': '@nativescript/core'
    },
    extensions: [
      '.ios.svelte',
      '.svelte',
      '.ios.ts',
      '.ts',
      '.ios.js',
      '.js',
      '.ios.mjs',
      '.mjs',
      '.ios.css',
      '.css',
      '.ios.scss',
      '.scss',
      '.ios.json',
      '.json'
    ],
    modules: [
      '__jest__/node_modules',
      'node_modules'
    ]
  },
  resolveLoader: {
    modules: [
      '__jest__/node_modules/@nativescript/webpack/dist/loaders',
      '__jest__/node_modules',
      'node_modules'
    ]
  },
  module: {
    rules: [
      /* config.module.rule('bundle') */
      {
        enforce: 'post',
        test: '__jest__/src/app.js',
        use: [
          /* config.module.rule('bundle').use('app-css-loader') */
          {
            loader: 'app-css-loader',
            options: {
              platform: 'ios'
            }
          }
        ]
      },
      /* config.module.rule('workers') */
      {
        test: /\\.(js|ts|svelte)$/,
        use: [
          /* config.module.rule('workers').use('nativescript-worker-loader') */
          {
            loader: 'nativescript-worker-loader'
          }
        ]
      },
      /* config.module.rule('ts') */
      {
        test: [
          /\\.ts$/
        ],
        use: [
          /* config.module.rule('ts').use('ts-loader') */
          {
            loader: 'ts-loader',
            options: {
              transpileOnly: true,
              allowTsInNodeModules: true,
              compilerOptions: {
                sourceMap: true,
                declaration: false
              },
              getCustomTransformers: function () { /* omitted long function */ }
            }
          }
        ]
      },
      /* config.module.rule('js') */
      {
        test: /\\.js$/,
        exclude: [
          /node_modules/
        ]
      },
      /* config.module.rule('xml') */
      {
        test: /\\.xml$/,
        use: [
          /* config.module.rule('xml').use('xml-namespace-loader') */
          {
            loader: 'xml-namespace-loader'
          }
        ]
      },
      /* config.module.rule('css') */
      {
        test: /\\.css$/,
        use: [
          /* config.module.rule('css').use('apply-css-loader') */
          {
            loader: 'apply-css-loader'
          },
          /* config.module.rule('css').use('css2json-loader') */
          {
            loader: 'css2json-loader'
          },
          /* config.module.rule('css').use('postcss-loader') */
          {
            loader: 'postcss-loader',
            options: {
              postcssOptions: {
                plugins: [
                  [
                    'postcss-import',
                    {
                      resolve: function () { /* omitted long function */ }
                    }
                  ]
                ]
              }
            }
          }
        ]
      },
      /* config.module.rule('scss') */
      {
        test: /\\.scss$/,
        use: [
          /* config.module.rule('scss').use('apply-css-loader') */
          {
            loader: 'apply-css-loader'
          },
          /* config.module.rule('scss').use('css2json-loader') */
          {
            loader: 'css2json-loader'
          },
          /* config.module.rule('scss').use('postcss-loader') */
          {
            loader: 'postcss-loader',
            options: {
              postcssOptions: {
                plugins: [
                  [
                    'postcss-import',
                    {
                      resolve: function () { /* omitted long function */ }
                    }
                  ]
                ]
              }
            }
          },
          /* config.module.rule('scss').use('sass-loader') */
          {
            loader: 'sass-loader'
          }
        ]
      },
      /* config.module.rule('svelte') */
      {
        test: /\\.svelte$/,
        exclude: [
          /node_modules/
        ],
        use: [
          /* config.module.rule('svelte').use('svelte-loader-hot') */
          {
            loader: 'svelte-loader-hot',
            options: {
              dev: true,
              preprocess: undefined,
              hotReload: true,
              hotOptions: {
                injectCss: false,
                'native': true
              },
              onwarn: function () { /* omitted long function */ }
            }
          }
        ]
      }
    ]
  },
  optimization: {
    runtimeChunk: 'single',
    splitChunks: {
      cacheGroups: {
        defaultVendor: {
          test: /[\\\\/]node_modules[\\\\/]/,
          priority: -10,
          name: 'vendor',
          chunks: 'all'
        }
      }
    },
    minimizer: [
      /* config.optimization.minimizer('TerserPlugin') */
      new TerserPlugin(
        {
          terserOptions: {
            compress: {
              collapse_vars: true,
              sequences: true,
              keep_infinity: true,
              drop_console: false,
              global_defs: {
                __UGLIFIED__: true
              }
            },
            keep_fnames: true,
            keep_classnames: true,
            format: {
              keep_quoted_props: true
            }
          }
        }
      )
    ]
  },
  plugins: [
    /* config.plugin('ForkTsCheckerWebpackPlugin') */
    new ForkTsCheckerWebpackPlugin(
      {
        async: false,
        typescript: {
          memoryLimit: 4096
        }
      }
    ),
    /* config.plugin('PlatformSuffixPlugin') */
    new PlatformSuffixPlugin(
      {
        extensions: [
          'ios'
        ]
      }
    ),
    /* config.plugin('ContextExclusionPlugin|App_Resources') */
    new ContextExclusionPlugin(
      /(.*)App_Resources(.*)/
    ),
    /* config.plugin('ContextExclusionPlugin|Other_Platforms') */
    new ContextExclusionPlugin(
      /\\.(android|visionos)\\.(\\w+)$/
    ),
    /* config.plugin('DefinePlugin') */
    new DefinePlugin(
      {
        __DEV__: true,
        __NS_WEBPACK__: true,
        __NS_ENV_VERBOSE__: false,
        __NS_DEV_HOST_IPS__: '["127.0.0.1","192.168.0.10"]',
        __CSS_PARSER__: '"css-tree"',
        __UI_USE_XML_PARSER__: true,
        __UI_USE_EXTERNAL_RENDERER__: false,
        __ANDROID__: false,
        __IOS__: true,
        __VISIONOS__: false,
        __APPLE__: true,
        'global.isAndroid': false,
        'global.isIOS': true,
        'global.isVisionOS': false,
        process: 'global.process'
      }
    ),
    /* config.plugin('CopyWebpackPlugin') */
    new CopyPlugin(
      {
        patterns: [
          {
            from: 'assets/**',
            context: '__jest__/src',
            noErrorOnMissing: true,
            globOptions: {
              dot: false,
              ignore: []
            }
          },
          {
            from: 'fonts/**',
            context: '__jest__/src',
            noErrorOnMissing: true,
            globOptions: {
              dot: false,
              ignore: []
            }
          },
          {
            from: '**/*.+(jpg|png)',
            context: '__jest__/src',
            noErrorOnMissing: true,
            globOptions: {
              dot: false,
              ignore: []
            }
          }
        ]
      }
    ),
    /* config.plugin('WatchStatePlugin') */
    new WatchStatePlugin()
  ],
  entry: {
    bundle: [
      '@nativescript/core/globals/index',
      '@nativescript/core/bundle-entry-points',
      '__jest__/src/app.js'
    ],
    'tns_modules/inspector_modules': [
      '@nativescript/core/inspector_modules'
    ]
  }
}"
`;
